File: Data-Structures-and-Algorithms/0344-ReverseString/reverseString.py
class Mutations:
    def reverseString(self, s):
        left, right = 0, len(s) - 1
        while left < right:
            s[left], s[right] = s[right], s[left]
            left, right = left + 1, right - 1
        return s

arr = ["h","e","l","l","o"]
print(arr)

mutations = Mutations()
mutations.reverseString(arr)
print(arr)

# O(N) for time (N/2 swaps)
# O(1) for space (no recursion)
# Two pointer, left end and right end, move inwards.

File: Data-Structures-and-Algorithms/0344-ReverseString/reverseString.js
var arr = ["h","e","l","l","o"];
console.log(arr);

const reverseString = function(s) {
    let left = 0;
    let right = s.length - 1;
    
    while(right > left) {
      let temp =  s[right];
        s[right--] = s[left];
        s[left++] = temp;
    }
    return s
};

console.log(reverseString(arr));
console.log(arr);

// O(N) for time (N/2 swaps)
// O(1) for space (no recursion)
// Two pointer, left end and right end, move inwards.